src/Errors.sol

```diff
@@ -0,0 +1,295 @@
+
+    /// @notice Cannot increase the payment rate or change the rate on a terminated rail
+    /// @param railId The ID of the rail
+    error RateChangeNotAllowedOnTerminatedRail(uint256 railId);
+
+    /// @notice Account lockup must be fully settled to change the payment rate on an active rail
+    /// @param railId The ID of the rail
+    /// @param from The address whose lockup is being checked
+    /// @param isSettled Whether the account lockup is fully settled
+    /// @param currentRate The current payment rate
+    /// @param attemptedRate The attempted new payment rate
+    error LockupNotSettledRateChangeNotAllowed(
+        uint256 railId, address from, bool isSettled, uint256 currentRate, uint256 attemptedRate
+    );
+
+    /// @notice Payer's lockup rate is less than the old payment rate when updating an active rail
+    /// @param railId The ID of the rail
+    /// @param from The address whose lockup is being checked
+    /// @param lockupRate The current lockup rate of the payer
+    /// @param oldRate The current payment rate for the rail
+    error LockupRateLessThanOldRate(uint256 railId, address from, uint256 lockupRate, uint256 oldRate);
+
+    /// @notice The payer does not have enough funds for the one-time payment
+    /// @param token The token being used for payment
+    /// @param from The payer's address
+    /// @param required The amount required (oneTimePayment)
+    /// @param actual The actual funds available in the payer's account
+    error InsufficientFundsForOneTimePayment(address token, address from, uint256 required, uint256 actual);
+
+    /// @notice Cannot settle a terminated rail without validation until after the max settlement epoch has passed
+    /// @param railId The ID of the rail being settled
+    /// @param currentBlock The current block number (actual)
+    /// @param requiredBlock The max settlement epoch block (expected, must be exceeded)
+    error CannotSettleTerminatedRailBeforeMaxEpoch(
+        uint256 railId,
+        uint256 requiredBlock, // expected (maxSettleEpoch + 1)
+        uint256 currentBlock // actual (block.number)
+    );
+
+    /// @notice Cannot settle a rail for epochs in the future.
+    /// @param railId The ID of the rail being settled
+    /// @param maxAllowedEpoch The latest epoch that can be settled (expected, must be >= actual)
+    /// @param attemptedEpoch The epoch up to which settlement was attempted (actual)
+    error CannotSettleFutureEpochs(uint256 railId, uint256 maxAllowedEpoch, uint256 attemptedEpoch);
+
+    /// @notice No progress was made in settlement; settledUpTo did not advance.
+    /// @param railId The ID of the rail
+    /// @param expectedSettledUpTo The expected value for settledUpTo (must be > startEpoch)
+    /// @param actualSettledUpTo The actual value after settlement attempt
+    error NoProgressInSettlement(uint256 railId, uint256 expectedSettledUpTo, uint256 actualSettledUpTo);
```

Comment:
nit: it might be useful to take invariant violations like this and give their errors a common prefix so we can quickly look at errors and know if they represent programmer errors.  i.e. `InvariantViolationNoProgressInSettlement` or `InternalErrNoProgressInSettlement`.  Just an idea.  It would be a decent amount of more work because you'd have to go through and understand which errors represent valid problems related to user input and which errors represent the system being in a bad and unaccounted for state.

Reply:
Some other things that I think meet this criterion: `InvalidRateChangeQueueState`, `LockupInconsistencyDuringRailFinalization`, `LockupExceedsFundsInvariant`